File: verses/VersesTableViewController.m
#import "verses-Swift.h"
#import "VersesTableViewController.h"
#import "AddVerseViewController.h"
#import "VerseDetailTableViewController.h"

@interface VersesTableViewController ()
@property (nonatomic, strong) NSFetchedResultsController *fetchedResultsController;
@property (nonatomic, strong) NSIndexPath *disclosingRowIndexPath;
@property (nonatomic, strong) NSManagedObjectContext *userManagedObjectContext;
@end

@implementation VersesTableViewController

@synthesize fetchedResultsController=_fetchedResultsController;

-(void)viewDidLoad {
    self.tableView.allowsMultipleSelectionDuringEditing = NO;
    
    AppDelegate *appDelegate = [[UIApplication sharedApplication] delegate];
    self.userManagedObjectContext = appDelegate.managedObjectContext;
    
    NSError *error;
    if (![[self fetchedResultsController] performFetch:&error]) {
      NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
      abort();
    }
}

#pragma mark - Fetched results controller

- (NSFetchedResultsController *)fetchedResultsController {
  if (_fetchedResultsController != nil) {
    return _fetchedResultsController;
  }
  
  NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] init];
  NSEntityDescription *entity = [NSEntityDescription entityForName:@"BiblePassage" inManagedObjectContext:self.userManagedObjectContext];
  [fetchRequest setEntity:entity];
  
  NSSortDescriptor *numberDescriptor = [[NSSortDescriptor alloc] initWithKey:@"passage" ascending:YES];
  NSArray *sortDescriptors = @[numberDescriptor];
  [fetchRequest setSortDescriptors:sortDescriptors];

  _fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:self.userManagedObjectContext sectionNameKeyPath:nil cacheName:nil];
  _fetchedResultsController.delegate = self;
  
  return _fetchedResultsController;
}

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
  [self.tableView beginUpdates];
}

- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type newIndexPath:(NSIndexPath *)newIndexPath {
  
  UITableView *tableView = self.tableView;
  
  switch(type) {
    case NSFetchedResultsChangeInsert:
      [tableView insertRowsAtIndexPaths:@[newIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
      break;
      
    case NSFetchedResultsChangeDelete:
      [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
      break;
    case NSFetchedResultsChangeMove:
      break;
    case NSFetchedResultsChangeUpdate:
      break;
  }
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
  [self.tableView endUpdates];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return [[self.fetchedResultsController sections] count];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
  return [sectionInfo numberOfObjects];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  BiblePassage *passage = [self.fetchedResultsController objectAtIndexPath:indexPath];
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"verseCell" forIndexPath:indexPath];
  UILabel *titleLabel = (UILabel *)[cell viewWithTag:0];
  titleLabel.text = passage.passage;
  return cell;
}

-(void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
  if (editingStyle == UITableViewCellEditingStyleDelete) {
    BiblePassage *passage = [self.fetchedResultsController objectAtIndexPath:indexPath];
    [self.userManagedObjectContext deleteObject:passage];

    NSError *error;
    [self.userManagedObjectContext save:&error];
  }
}
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  self.disclosingRowIndexPath = indexPath;
  [self performSegueWithIdentifier:@"verseDetail" sender:self];
}

#pragma mark - Navigation

-(IBAction)unwindToList:(UIStoryboardSegue *)segue {
  
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
  if ([segue.identifier isEqualToString:@"addVerse"]) {
    AddVerseViewController *addVerseController = [segue destinationViewController];
    addVerseController.bibleAPI = [[BibliaAPI alloc] initWithMoc:self.userManagedObjectContext];
  }
  
  if ([segue.identifier isEqualToString:@"verseDetail"]) {
    VerseDetailTableViewController *detailController = [segue destinationViewController];
    detailController.biblePassage = [self.fetchedResultsController objectAtIndexPath:self.disclosingRowIndexPath];
  }
}

@end

File: verses/VersesTableViewController.swift
//
//  VersesTableViewController.swift
//  verses
//
//  Created by Isaac Williams on 1/31/15.
//  Copyright (c) 2015 The Williams Family. All rights reserved.
//

import Foundation
import UIKit
import CoreData

class VersesTableViewController: UITableViewController, NSFetchedResultsControllerDelegate, AddVerseDelegate {
    // MARK: Variable and constant declaration
    
    let appDelegate: AppDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
    var managedObjectContext: NSManagedObjectContext?
    var _orderedPassages: [BiblePassage]?

    lazy var fetchedResultsController: NSFetchedResultsController = {
        let fetchRequest = NSFetchRequest()
        let entity = NSEntityDescription.entityForName("BiblePassage", inManagedObjectContext: self.managedObjectContext!)
        fetchRequest.entity = entity!

        let numberDescriptor = NSSortDescriptor(key: "passage", ascending: true)
        let sortDescriptors = [numberDescriptor]
        fetchRequest.sortDescriptors = sortDescriptors

        let fetchedResultsController = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: self.managedObjectContext!, sectionNameKeyPath: nil, cacheName: nil)
        fetchedResultsController.delegate = self

        return fetchedResultsController
    }()
    
    // MARK: Setup

    override func viewDidLoad() {
        self.tableView.allowsMultipleSelectionDuringEditing = false
        self.managedObjectContext = self.appDelegate.managedObjectContext
        self.fetchedResultsController.performFetch(nil)
    }
    
    // MARK: Custom methods
    
    func orderedPassages() -> [BiblePassage] {
        if _orderedPassages == nil { _orderedPassages = self.fetchedResultsController.fetchedObjects as? [BiblePassage] }
        let activePassage = self.appDelegate.biblePassageStore.activeBiblePassage
        
        if activePassage == nil { return _orderedPassages! }
        
        let otherPassages = _orderedPassages!.filter {$0 != activePassage}
        _orderedPassages = [activePassage!] + otherPassages
        
        return _orderedPassages!
    }
    
    func flagCell(cell: UITableViewCell, atIndexPath indexPath: NSIndexPath) {
        let activePassage = orderedPassages()[indexPath.row] == self.appDelegate.biblePassageStore.activeBiblePassage
        cell.accessoryView = activePassage ? UIImageView(image: UIImage(named: "flag.png")) : nil
    }
    
    // MARK: Table View methods

    func controllerWillChangeContent(controller: NSFetchedResultsController) {
        self.tableView.beginUpdates()
    }

    func controller(controller: NSFetchedResultsController, didChangeObject anObject: AnyObject, atIndexPath indexPath: NSIndexPath?, forChangeType type: NSFetchedResultsChangeType, newIndexPath: NSIndexPath?) {
        let tableView: UITableView = self.tableView

        switch type {
        case NSFetchedResultsChangeType.Insert:
            tableView.insertRowsAtIndexPaths([newIndexPath!], withRowAnimation: UITableViewRowAnimation.Automatic)
        case NSFetchedResultsChangeType.Delete:
            tableView.deleteRowsAtIndexPaths([indexPath!], withRowAnimation: UITableViewRowAnimation.Automatic)
        default:
            break
        }
    }
    
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        switch editingStyle {
        case .Delete:
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: UITableViewRowAnimation.Automatic)
        default:
            break
        }
    }

    func controllerDidChangeContent(controller: NSFetchedResultsController) {
        self.tableView.endUpdates()
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return self.fetchedResultsController.sections!.count
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo: NSFetchedResultsSectionInfo = self.fetchedResultsController.sections![section] as! NSFetchedResultsSectionInfo
        return sectionInfo.numberOfObjects
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let passage: BiblePassage = self.orderedPassages()[indexPath.row]
        let cell: UITableViewCell = tableView.dequeueReusableCellWithIdentifier("verseCell", forIndexPath: indexPath) as! UITableViewCell
        if passage.passage != nil { cell.textLabel!.text = passage.passage }
//        flagCell(cell, atIndexPath: indexPath)
        
        return cell
    }
    
    override func tableView(tableView: UITableView, editActionsForRowAtIndexPath indexPath: NSIndexPath) -> [AnyObject]? {
        let passage = self.orderedPassages()[indexPath.row]
        var actions = [UITableViewRowAction]()
        
        let deleteAction = UITableViewRowAction(style: .Normal, title: "Delete", handler: { (action: UITableViewRowAction!, indexPath: NSIndexPath!) in
            self.managedObjectContext?.deleteObject(passage)
            self.managedObjectContext?.save(nil)
        })
        
        deleteAction.backgroundColor = UIColor.redColor()
        actions.append(deleteAction)

        if orderedPassages()[indexPath.row] != self.appDelegate.biblePassageStore.activeBiblePassage {
            let flagAction = UITableViewRowAction(style: .Normal, title: "Flag", handler: { (action: UITableViewRowAction!, indexPath: NSIndexPath!) in
                self.appDelegate.exportToTodayApp(passage)
                self.appDelegate.biblePassageStore.activeBiblePassage = self.orderedPassages()[indexPath.row]
                
                let topIndexPath = NSIndexPath(forRow: 0, inSection: 0)
                //            tableView.reloadRowsAtIndexPaths([indexPath, topIndexPath], withRowAnimation: .None)
                
                tableView.moveRowAtIndexPath(indexPath, toIndexPath: topIndexPath)
                tableView.editing = false
            })
            flagAction.backgroundColor = UIColor(red: 1.0, green: 0.6, blue: 0.0, alpha: 1.0)
            actions.append(flagAction)
        }
        
        return actions
    }
    
    // MARK: Segue Control

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "verseDetail" {
            let detailController: VerseDetailTableViewController = segue.destinationViewController as! VerseDetailTableViewController
            detailController.biblePassage = self.orderedPassages()[self.tableView.indexPathForSelectedRow()!.row]
        }
    }
    
    func addVerseCanceled() {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func verseAdded(passage: BiblePassage) {
        self.appDelegate.biblePassageStore.activeBiblePassage = passage
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    @IBAction func addVerse(sender: AnyObject) {
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        let addVerseNavigationController = storyboard.instantiateViewControllerWithIdentifier("addVerse") as! UINavigationController
        addVerseNavigationController.modalTransitionStyle = .CoverVertical
        let addVerseController = addVerseNavigationController.childViewControllers[0] as! AddVerseViewController
        addVerseController.delegate = self
        self.presentViewController(addVerseNavigationController, animated: true, completion: nil)
    }
    
}